### ISSUE TITLE
Customize search element

### ISSUE THREAD

**Issue by NickToye** [gregnb/mui-datatables#91]
Expose search component so that it can be customized, such as adding a Placeholder.

**Comment by NickToye** [comment]
Good news, although you've spelt Customise wrong. ;)

**Comment by gregnb** [comment]
Really? http://www.dictionary.com/browse/customize

**Comment by NickToye** [comment]
Americanisms.

**Comment by chiran-subedi** [comment]
Hello Gregnb,

I hope you can help me with this. I have been using mui-datatable and its
too good. Firstly i'd like to thankyou for this awesome project.

I want some customization on the tooltips.

Right now, whenever i hover through each of the column head at the table,
the tooltip popover shows "sort" for all. Is there any way so that we can
provide different titles for each of the columns ?
Expected Behavior

const options = {
textLabels: {
body: {
toolTip: "sort"
}
}

There is an element for this inside textLabels which is toolTip, but this
shows for all of the column's head. Can you please help me with having this
functionality(tooltip popover) work for each column heads with different
strings ?

I have a screenshot for more clarification.

Thankyou in advance. :)
Chiran

**Comment by gregnb** [comment]
@chiran-subedi Thanks for the feedback, but please file an issue. Posting your own issues in other people's issues can be inconsiderate. So if you can please remove this comment and the other and file an issue. Thank you

**Comment by gregnb** [comment]
So #96 has a similar question about more customization support, but for the column header. I have given it some time and I'm trying to come up with a way to solve the current issues and get ahead of future issues.  I think it's going to require some breaking changes and what I'm proposing is the following changes to options:

Change `options` object on custom columns from `customRender` to `customBodyCell` which will then setup nicely `customHeadCell`:

```
const columns = [
      {
        name: "Active",
        options: {
          filter: true,
          customHeadCell: {
             render: (value, tableMeta, updateValue) => {}
          },
          customBodyCell: {
            render: (value, tableMeta, updateValue) => {
              return (
              <FormControlLabel
                label={value ? "Yes" : "No"}
                value={value ? "Yes" : "No"}
                control={
                  <Switch color="primary" checked={value} value={value ? "Yes" : "No"} />
                }
                onChange={event => {
                  updateValue(event.target.value === "Yes" ? false : true);
                }}
              />
            );

          }
        }
      }
    ];

```

and then on the `options` prop adding a property named `exposeInternals` or `exposeMUI` (I need a better name here) that would expose the underlying Material-UI components on different sections in the table. That way, if someone has a request like you do above you can go ahead and override without me having to solve requests for opening such access on a ticket by ticket basis.  

```
const options = {
  exposeMUI: {
    search: {
       textField: { someProp: true }
    },
    ...
  }
}; 
```

then you could go ahead and add a placeholder via that reference

**Comment by NickToye** [comment]
Happy to help.  I’m not at my computer at the moment, but I’ll have a think.

**Comment by allenjoy123** [comment]
Has this feature been implemented yet. I'd like to be able to allow user to search by entering search terms so as not generate a backend request for every letter.

**Comment by allenjoy123** [comment]
I have a requirement to have a search button so that the user can trigger and wait for the search request to complete. 
I have used lodash debounce to trigger the search but since the response needs to come from the server, it is easy to trigger multiple search requests to the server that overlap each other and confuse the user.
I'd like to be able to add a Search chip beside the 'Close' chip that occurs on clicking on the search icon. Can this component be exposed so I can customize it.

**Comment by somsub** [comment]
@gregnb Thanks for this awesome component. We also have a requirement to customize the search input control. If there is any assistance I can provide please let me know. `exposeMUI` seems be a good name if you want to proceed with that. If you don't have bandwidth, I can create a PR with some guidance from you. Thanks

**Comment by somsub** [comment]
@gregnb Can we do something where the Implementing App (IA) can pass in a Custom Search component that is similar to what we have 
```<CustomTableSearch onSearch={...} onHide={...} options={options} />```. If the IA passes one, then mount that, else mount the inbuilt **TableSearch**. 

Thoughts?

**Comment by gabrielliwerant** [comment]
The ability to add a custom search bar is now available in `2.8.0`. See documentation or release notes for more information.
